lib/roby/app.rb

```diff
@@ -2997,11 +2997,18 @@ def remove_notification_listener(listener)
         # @param [Boolean] only_self if set, list only test files from within
         #   {#app_dir}. Otherwise, consider test files from all over {#search_path}
         # @return [Array<String>]
-        def discover_test_files(all: true, only_self: false)
+        def discover_test_files(all: true, only_self: false, base_dir: File.join(app_dir, "test"))
             if all
-                test_files = each_test_file_in_app.each_with_object({}) do |k, h|
+                robot_test_files = each_test_file_in_robot.each_with_object({}) do |k, h|
                     h[k] = []
                 end
+                base_dir_test_files = each_test_file_in(base_dir)
+                base_dir_test_files = base_dir_test_files.each(base_dir).with_object({}) do |k, h|
```

Comment:
You are again encoding a behavior that is very specific to our payload structure, plus, you would not run a test/robots/test_ROBOT.rb file from the payload itself.

Instead of doing that, we need a way to run the tests that the payload needs from the CLI interface, e.g.

~~~
roby test --base-dir $BASE_DIR test/robots/test_ROBOT.rb [other options]
~~~

It does mean you have to change scripts/tests.rb further to:
- have an option that says "call discover_test_files even if there are explicit files on the command line" (e.g. `--discover` or `--auto`). Right now, passing test/robots/test_ROBOT.rb would disable discovery.
- make the test script replace ROBOT with the robot name or category, and pick the one(s) that exist.

Reply:
`--force-discovery`?

Reply:
I'm thinking on that:

```rb
if test_files.empty?
    test_files = discover_test_files
    if list_tests
        puts "Would load #{test_files.size} test files"
        test_files.each do |path|
            puts "  #{path}"
        end

        all_existing_tests = app.find_dirs("test", order: :specific_first, all: !only_self).inject(Set.new) do |all, dir|
            all.merge(Find.enum_for(:find, dir).find_all { |f| f =~ /\/test_.*\.rb$/ && File.file?(f) }.to_set)
        end
        not_run = (all_existing_tests - test_files.to_set)
        unless not_run.empty?
            puts "\nWould NOT load #{not_run.size} tests"
            not_run.to_a.sort.each do |not_loaded|
                puts "  #{not_loaded}"
            end
        end
        exit 0
    end
elsif force_discovery
    test_files = discover_test_files

    test_files.map do |file|
        next file unless file =~ /ROBOT/

        (replacements = [robot_type]) << robot_name if robot_type != robot_name
        replacements.flat_map do |replacement|
            file.gsub("ROBOT", replacement)
        end
    end
end
```

Reply:
Looks good, just
- make the list case a common case (since we can discover and replace ROBOT)
- when replacing, you need to remove files that do not exist

Reply:
> make the list case a common case (since we can discover and replace ROBOT)

I didn't quite get this specific part

Reply:
unless you mean something like:

```rb
        (test_files = discover_test_files(all: all, only_self: only_self, base_dir: base_dir)) if test_files.empty? || force_discovery
        if test_files.empty?
            if list_tests
                puts "Would load #{test_files.size} test files"
                test_files.each do |path|
                    puts "  #{path}"
                end

                all_existing_tests = Roby.app.find_dirs("test", order: :specific_first, all: !only_self).inject(Set.new) do |all, dir|
                    all.merge(Find.enum_for(:find, dir).find_all { |f| f =~ /\/test_.*\.rb$/ && File.file?(f) }.to_set)
                end
                not_run = (all_existing_tests - test_files.to_set)
                unless not_run.empty?
                    puts "\nWould NOT load #{not_run.size} tests"
                    not_run.to_a.sort.each do |not_loaded|
                        puts "  #{not_loaded}"
                    end
                end
                exit 0
            end
        elsif force_discovery
            test_files = test_files.flat_map do |file|
                next file unless file =~ /ROBOT/

                robot_type = Roby.app.robot_type
                robot_name = Roby.app.robot_name
                (replacements = [robot_type]) << robot_name if robot_type != robot_name
                replacements.map do |replacement|
                    resolved_file = File.join(Roby.app.app_dir, file.gsub("ROBOT", replacement))
                    resolved_file if File.file?(resolved_file)
                end.compact
            end
        end
```

Reply:
[EDIT]
Just realized that you meant:

```rb
if test_files.empty?
            test_files = discover_test_files(all: all, only_self: only_self, base_dir: base_dir)
        elsif force_discovery
            test_files += discover_test_files(all: all, only_self: only_self, base_dir: base_dir)

            test_files = test_files.flat_map do |file|
                next file unless file =~ /ROBOT/

                robot_type = Roby.app.robot_type
                robot_name = Roby.app.robot_name
                (replacements = [robot_type]) << robot_name if robot_type != robot_name
                replacements.map do |replacement|
                    resolved_file = File.join(Roby.app.app_dir, file.gsub("ROBOT", replacement))
                    resolved_file if File.file?(resolved_file)
                end.compact
            end
        end

        if list_tests
            puts "Would load #{test_files.size} test files"
            test_files.each do |path|
                puts "  #{path}"
            end

            all_existing_tests = Roby.app.find_dirs("test", order: :specific_first, all: !only_self).inject(Set.new) do |all, dir|
                all.merge(Find.enum_for(:find, dir).find_all { |f| f =~ /\/test_.*\.rb$/ && File.file?(f) }.to_set)
            end
            not_run = (all_existing_tests - test_files.to_set)
            unless not_run.empty?
                puts "\nWould NOT load #{not_run.size} tests"
                not_run.to_a.sort.each do |not_loaded|
                    puts "  #{not_loaded}"
                end
            end
            exit 0
        end
```

Reply:
done